src/routers/const_router.rs

```diff
@@ -1,120 +1,71 @@
 use std::sync::RwLock;
 // pyo3 modules
 use crate::executors::execute_function;
+use anyhow::Context;
 use log::debug;
 use pyo3::prelude::*;
 use pyo3::types::PyAny;
 
 use actix_web::http::Method;
-use matchit::Router;
 
-use anyhow::{bail, Error, Result};
+use anyhow::{Error, Result};
 
-/// Contains the thread safe hashmaps of different routes
+use super::RouteType;
+use super::Router;
+
+type RouteMap = RwLock<matchit::Router<String>>;
 
+/// Contains the thread safe hashmaps of different routes
 pub struct ConstRouter {
-    get_routes: Arc<RwLock<Router<String>>>,
-    post_routes: Arc<RwLock<Router<String>>>,
-    put_routes: Arc<RwLock<Router<String>>>,
-    delete_routes: Arc<RwLock<Router<String>>>,
-    patch_routes: Arc<RwLock<Router<String>>>,
-    head_routes: Arc<RwLock<Router<String>>>,
-    options_routes: Arc<RwLock<Router<String>>>,
-    connect_routes: Arc<RwLock<Router<String>>>,
-    trace_routes: Arc<RwLock<Router<String>>>,
+    routes: HashMap<Method, Arc<RouteMap>>,
 }
 
-impl ConstRouter {
-    pub fn new() -> Self {
-        Self {
-            get_routes: Arc::new(RwLock::new(Router::new())),
-            post_routes: Arc::new(RwLock::new(Router::new())),
-            put_routes: Arc::new(RwLock::new(Router::new())),
-            delete_routes: Arc::new(RwLock::new(Router::new())),
-            patch_routes: Arc::new(RwLock::new(Router::new())),
-            head_routes: Arc::new(RwLock::new(Router::new())),
-            options_routes: Arc::new(RwLock::new(Router::new())),
-            connect_routes: Arc::new(RwLock::new(Router::new())),
-            trace_routes: Arc::new(RwLock::new(Router::new())),
-        }
-    }
-
-    #[inline]
-    fn get_relevant_map(&self, route: Method) -> Option<Arc<RwLock<Router<String>>>> {
-        match route {
-            Method::GET => Some(self.get_routes.clone()),
-            Method::POST => Some(self.post_routes.clone()),
-            Method::PUT => Some(self.put_routes.clone()),
-            Method::PATCH => Some(self.patch_routes.clone()),
-            Method::DELETE => Some(self.delete_routes.clone()),
-            Method::HEAD => Some(self.head_routes.clone()),
-            Method::OPTIONS => Some(self.options_routes.clone()),
-            Method::CONNECT => Some(self.connect_routes.clone()),
-            Method::TRACE => Some(self.trace_routes.clone()),
-            _ => None,
-        }
-    }
-
-    #[inline]
-    fn get_relevant_map_str(&self, route: &str) -> Option<Arc<RwLock<Router<String>>>> {
-        if route != "WS" {
-            let method = match Method::from_bytes(route.as_bytes()) {
-                Ok(res) => res,
-                Err(_) => return None,
-            };
-
-            self.get_relevant_map(method)
-        } else {
-            None
-        }
-    }
-
-    /// Checks if the functions is an async function
-    /// Inserts them in the router according to their nature(CoRoutine/SyncFunction)
+impl Router<String, Method> for ConstRouter {
```

Comment:
Hey @AntoineRR ,

I am bit unfamiliar with `traits` in rust. Can you please explain this statement to me. And the benefits associate with it?

Reply:
Sure !

## About traits

Traits are a way in Rust to share behavior between structs. There is no inheritance in Rust as in a truly object oriented language like C++, so traits are more similar to interfaces in Java for example.

Here, defining the `Router` trait allows to specify what methods should be in a router : `add_route` and `get_route`. For every router we want to define, we have to implement this trait using the line above. This will force us to provide a definition for the `add_route` and `get_route` method that is specific to our struct, which is `ConstRouter` here.

## About the generics

I didn't want to change the code too much so I kept the same return types and parameter types in every router struct. The return type of the `get_route` method is different across the routers (`DynRouter` and `MiddlewareRouter` return an `Option<((PyFunction, u8), HashMap<String, String>)>` while `ConstRouter` return an `Option<String>`), so I used a generic type to unify their behavior in the trait. Similarly, the `route_type` parameter of the `get_route` method can either be a `MiddlewareRoute` or a `Method`. I needed two different generic types to handle those differences: this is `T` and `U` in the definition of the trait : `pub trait Router<T, U>`. For `ConstRouter`, `T` is `String` and `U` is `Method`.

## Benefits of using traits

With this PR, `DynRouter`, `ConstRouter` and `MiddlewareRouter` all implement the `Router` trait (I wanted to include the `WebSocketRouter` too but it seems too different from the others). This means the compiler is aware those share some behavior. We could now do some cool things like:
- Creating a `Vec<dyn Router>` containing the routers
- Create generic methods that take a type implementing the `Router` trait as parameter -> `fn method<T: Router>(router: T)`
- Eventually provide a default implementation for the `add_route` and `get_route` methods if we can make it the same for all structs (which I may try to do later)
- If a new method needs to be added for every router, its implementation can be in the `Router` trait instead of in each struct
- Maybe other things I forgot?

I hope this makes things clearer! In this PR I didn't take advantage of the trait for now, I wanted to do small iterations on the code instead of a really big one. I will try to improve things further if I have some time :wink:

Reply:
Thank you @AntoineRR . This an excellent explanation 😄

Reply:
You're welcome ! 😉